<a href="https://colab.research.google.com/github/kavishbhardwaj/Rep1/blob/main/SC_Week0_Unit3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Week 0 - Unit 3: In-Depth `ls`, Hard and Soft Links, Inodes, and File System Size

## Introduction

In this unit, we will delve into the `ls` command, explore hard and soft links, understand inodes and their exploration, and check file system sizes.

## `ls` command

The `ls` command lists directory contents. It has many options to customize the output.

*   `-l`: Long format listing.
*   `-a`: List all files, including hidden files (those starting with a dot .).
*   `-h`: Human-readable file sizes (e.g., KB, MB).
*   `-R`: Recursively list subdirectories.
*   `-t`: Sort by modification time.


In [ ]:
# @title Example
%%bash
touch file1.txt
ls -l

total 4
-rw-r--r-- 1 root root    0 Sep  5 04:44 file1.txt
drwxr-xr-x 1 root root 4096 Sep  3 13:24 sample_data


**Detailed Explanation of `ls -l`:**

```
-rw-r--r-- 1 root root    0 Sep  5 04:44 file1.txt
```

1. `-rw-r--r--`: File permissions.
2. `1`: Number of links.
3. `root`: File owner.
4. `root`: File group.
5. `0`: File size in bytes.
6. `Sep 5 04:44`: Last modification date and time.
7. `file1.txt`: File name.



### Use Cases

1. Listing Files in Long Format: When you need detailed information about files (permissions, ownership, size, and modification date), use the `-l` option.

```
ls -l
```

This command provides a comprehensive view of the directory contents.

2. Including Hidden Files: To see all files, including hidden files (those starting with a dot), use the `-a` option.

```
ls -a
```

This is useful for checking configuration files or directories that are usually hidden.

3. Human-Readable File Sizes: When listing files, you may want to see sizes in a more understandable format. Combine `-l` with `-h` for human-readable sizes.

```
ls -lh
```

This command shows file sizes in KB, MB, etc., making it easier to assess storage usage.

4. Recursively Listing Subdirectories: If you want to list all files in the current directory and its subdirectories, use the `-R` option.

```
ls -R
```

This is particularly useful for exploring the entire directory structure of a project.

5. Sorting by Modification Time: To find out which files were recently modified, use the `-t` option.

```
ls -lt
```

This command sorts files by their modification time, displaying the most recently modified files at the top.

### Exercises

By practicing these exercises, you'll become more proficient in using the `ls` command

1. Basic Directory Listing:

* Open your terminal and run the command to list the files in your current directory.


In [ ]:
# @title Practice here
%%bash

UsageError: %%bash is a cell magic, but the cell body is empty.


2. Detailed Listing:

* List files in long format to see permissions and other details.

In [ ]:
# @title Practice here
%%bash

3. Hidden Files:

* Check for hidden files in your home directory. Note - You can directly use `~` to check your home directory.

In [ ]:
# @title Practice here
%%bash

4. Human-Readable Sizes:

* List files in your current directory with human-readable sizes.

In [ ]:
# @title Practice here
%%bash

5. Recursive Listing:

* Create a directory structure (e.g., `dir1/dir2/dir3`) and list all files recursively.

In [ ]:
# @title Practice here
%%bash

6. Sorting by Modification Time:

* Modify one of the files (e.g., `file1.txt`) and then run the command to sort files by modification time.

In [ ]:
# @title Practice here
%%bash

7. Combining Options:

* Use a combination of options to list all files, including hidden ones, in long format with human-readable sizes.

In [ ]:
# @title Practice here
%%bash

## Hard Links

How computers save data into hard drive....what are blocks....how a file system ....how the concept of i nodes arise....file descriptors....what is a directoryy

A hard link is an additional directory entry for an existing file. It points to the same inode as the original file, meaning both names reference the same data.



In [ ]:
# @title Creating a Hard Link:
%%bash
touch file1.txt
ln file1.txt file2.txt

`file1.txt` and `file2.txt` point to the same inode.

In [ ]:
# @title Checking Hard Links: Use `ls -li` to display inode numbers.
%%bash
ls -li

total 4
6946831 -rw-r--r-- 2 root root    0 Sep  5 05:05 file1.txt
6946831 -rw-r--r-- 2 root root    0 Sep  5 05:05 file2.txt
1572887 drwxr-xr-x 1 root root 4096 Sep  3 13:24 sample_data


You’ll see the same inode number (6946831) for both the original file and the hard link.

## Soft Links (Symbolic Links)

A soft link (or symbolic link) is a special type of file that refers to another file or directory by its path. explain more about paths....relative and absolute pathss...how to do it wrt ls and how same concepts can be used....snippet of such pathsss..

**Syntax**

```
ln -s /path/to/original_file.txt soft_link.txt
```

*   `-s` specifies a symbolic link.
*   `soft_link.txt` points to `/path/to/original_file.txt`.





## Understanding Inodes

An inode is a data structure on a filesystem that stores information about a file or directory (except its name or data).

permissions of hard links and soft linksss..can we modify the permissions of soft linkss

Inode Information Includes:

1. File Type
2. Permissions
3. Owner
4. Group
5. Size
6. Timestamps
7. Location of Data blocks



In [ ]:
# @title Viewing Inodes:
%%bash
ls -i

6946831 file1.txt
6946831 file2.txt
1572887 sample_data


This will display inode numbers alongside filenames.

### Examples

1. Exploring Inodes

* Create a new file and check its inode number.

In [ ]:
%%bash
touch myfile.txt
ls -i myfile.txt

5111817 myfile.txt


* Note down the inode number. Then, delete the file and create it again. Check the inode number again to see if it changes.

In [ ]:
%%bash
rm myfile.txt
touch myfile.txt
ls -i myfile.txt

5111817 myfile.txt


2. Creating Hard Links:

* Create a file named `original.txt` and add some content to it.

In [ ]:
%%bash
echo "This is the original file." > original.txt

* Create a hard link to this file named `hardlink.txt`.

In [ ]:
%%bash
ln original.txt hardlink.txt

* Verify that both files point to the same inode.

In [ ]:
%%bash
ls -i original.txt hardlink.txt

5111824 hardlink.txt
5111824 original.txt


* Now, modify the content of `hardlink.txt` and check if the change reflects in `original.txt`.

In [ ]:
%%bash
echo "Modified content." > hardlink.txt
cat original.txt

Modified content.


3. Creating Soft Links:

* Create a file named `source.txt` and write some text in it.

In [ ]:
%%bash
echo "This is the source file." > source.txt

* Create a symbolic link to this file named `symlink.txt`.

In [ ]:
%%bash
ln -s source.txt symlink.txt

* Check the details of the symbolic link using `ls -l`.

In [ ]:
%%bash
ls -l symlink.txt

lrwxrwxrwx 1 root root 10 Sep 18 10:06 symlink.txt -> source.txt


* Now, delete `source.txt` and check the content of `symlink.txt`.

In [ ]:
%%bash
cat symlink.txt

cat: symlink.txt: No such file or directory


CalledProcessError: Command 'b'cat symlink.txt\n'' returned non-zero exit status 1.

In [ ]:
# @title Exploring Inodes: To view detailed inode information:
%%bash
stat file1.txt

  File: file1.txt
  Size: 0         	Blocks: 0          IO Block: 4096   regular empty file
Device: 3dh/61d	Inode: 6946831     Links: 2
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2024-09-05 05:05:19.890859001 +0000
Modify: 2024-09-05 05:05:16.886596834 +0000
Change: 2024-09-05 05:05:16.891597271 +0000
 Birth: 2024-09-05 04:44:47.668325506 +0000


Displays inode number, file size, permissions, and much more.

## Inode Exploration

Each file has a unique inode number. To explore files and their inodes:

1. Create files and links.
2. Check inodes of these files.

In [ ]:
# @title Example
%%bash
touch file1.txt
ln file1.txt hardlink1.txt
ln -s file1.txt symlink1.txt
ls -li

total 4
6946831 -rw-r--r-- 3 root root    0 Sep  5 05:14 file1.txt
6946831 -rw-r--r-- 3 root root    0 Sep  5 05:14 file2.txt
6946831 -rw-r--r-- 3 root root    0 Sep  5 05:14 hardlink1.txt
1572887 drwxr-xr-x 1 root root 4096 Sep  3 13:24 sample_data
6946900 lrwxrwxrwx 1 root root    9 Sep  5 05:14 symlink1.txt -> file1.txt


Please note:

1. `file1.txt` and `hardlink1.txt` share the same inode number.
2. `symlink1.txt` has a different inode number, pointing to `file1.txt`.

## File System Size

File system size indicates the total space and usage on a filesystem. Use `df` and `du` commands for this purpose.

lsusp, lshw, lsmod, lsusb - i will use it later

In [ ]:
# @title Check Disk Space Usage:
%%bash
df -h

Filesystem      Size  Used Avail Use% Mounted on
overlay         108G   33G   75G  31% /
tmpfs            64M     0   64M   0% /dev
shm             5.8G     0  5.8G   0% /dev/shm
/dev/root       2.0G  1.2G  820M  59% /usr/sbin/docker-init
tmpfs           6.4G  2.0M  6.4G   1% /var/colab
/dev/sda1        70G   53G   18G  75% /etc/hosts
tmpfs           6.4G     0  6.4G   0% /proc/acpi
tmpfs           6.4G     0  6.4G   0% /proc/scsi
tmpfs           6.4G     0  6.4G   0% /sys/firmware


`-h` for human-readable format (e.g., GB, MB).

In [ ]:
# @title Check Directory Size:
%%bash
du -sh /home

4.0K	/home



*   `-s` for summary.
*   `-h` for human-readable format.



## Summary

In this unit, we explored the `ls` command in depth, learned about hard and soft links, understood inodes and how to explore them, and checked file system sizes. These concepts are fundamental for managing files and directories effectively in Linux.